src/language-js/needs-parens.js

```diff
@@ -687,6 +687,37 @@ function needsParens(path, options) {
     // fallthrough
     case "CallExpression":
     case "MemberExpression":
+      if (
+        path.match(
+          () =>
+            node.type === "CallExpression" || node.type === "MemberExpression",
+          (node, name) =>
+            name === "expression" && node.type === "ChainExpression"
+        ) &&
+        (path.match(
+          undefined,
+          undefined,
+          (node, name) =>
+            (name === "callee" &&
+              ((node.type === "CallExpression" && !node.optional) ||
+                node.type === "NewExpression")) ||
+            (name === "object" &&
+              node.type === "MemberExpression" &&
+              !node.optional)
+        ) ||
+          path.match(
+            undefined,
+            undefined,
+            (node, name) =>
+              name === "expression" && node.type === "TSNonNullExpression",
+            (node, name) =>
+              name === "object" && node.type === "MemberExpression"
+          ))
```

Comment:
It seems unreadable to me... Can you add comment or extract these conditions to variables?

Reply:
What kind of comment do you expect? Similar code exists

https://github.com/prettier/prettier/blob/67e121a20b9d5d983970bdc2dcba5de2bd0f7beb/src/language-js/print/function-parameters.js#L246
https://github.com/prettier/prettier/blob/67e121a20b9d5d983970bdc2dcba5de2bd0f7beb/src/language-js/print/object.js#L200

Reply:
I expect comments that describes what concrete syntax each matcher represents.

Reply:
[`e849c94`](https://github.com/prettier/prettier/pull/13735/commits/e849c94bbf45d4ceb502402fadf2ea425b5e4dc1)